### DISCUSSION TITLE
Post migration: Abuse and failure scenarios

### DISCUSSION THREAD

**Discussion by beholdnec** [mastodon/mastodon#22708] [General]
**Introduction**

![image](https://user-images.githubusercontent.com/9994394/209407479-34e886c0-75d9-4153-b345-341ed73af329.png)

Post migration (exporting toots from one instance and importing them into another) is an important feature that is not currently supported by Mastodon. While it is possible to migrate follows and followers, the toots will not be transferred, essentially causing a wipe and reset of one's account if migration if performed.

In this post, I will attempt to predict and discuss some abuse and failure scenarios that could be triggered by migration.

**1. An instance shuts down before toots can be exported.**
- Discussion: In order for toots to be migrated, the data will need to be transferred from one server to the other. This could be impossible if the old toots are not available.
- Mitigations: If a server is going to be shut down, the owners should announce the shutdown well in advance so that users can export their data. There should be a period before shutdown in which no more toots can be posted, and data can only be exported. Warnings and instructions should be deployed to all users of the instance. This is optimistic, I know.

**2. A malicious instance refuses to allow toots to be exported.**
_A malicious instance is one which has become untrustworthy, either by hackers or malicious owners._
- Discussion: A malicious instance could lock their users in by refusing to allow them to export their toots.
- Mitigations: ...

**3. A malicious instance allows toots to be exported, but the toots are modified or corrupted in some way.**
- Discussion: A malicious instance could troll their users by exporting fake toots during migration.
- Mitigations: ...

**4. A malicious instance owner sells accounts to scammers, who take control of the accounts and migrate them elsewhere.**
- Discussion: ...
- Mitigations: There should be a system where the Fediverse is alerted to such activity and implements countermeasures to prevent hacked accounts from being set up on their servers.

**5. A malicious user sets up an account and imports someone else's toots, fraudulently claiming to be another person.**
- Discussion: ...
- Mitigations: ...

**6. A user wants to migrate toots to a new server, but some of their toots violate the TOS of the new server.**
- Discussion: ...
- Mitigations: ...

**7. A user triggers a migration, then changes their mind and wants to cancel it.**
- Discussion: ...
- Mitigations: Presumably this will work similarly to the normal method of cancelling account migration.

**8. A user migrates their account to more than one other instances and/or accounts.**
- Discussion: Perhaps the user forgot about the initial migration, or perhaps they wish to "fork" their account into multiple new accounts.
- Mitigations: ...

**9. A user wishes to migrate only a portion of their toots over to a new instance.**
- Discussion: For example, they wish to keep only the last year of toots when they migrate.
- Mitigations: There should be options for selecting which toots to migrate.

**Comment by limdingwen** [comment]
Good discussion. Here's my contribution. I think this would be better as a wiki page, but whatever 😆 

2. A malicious instance refuses to allow toots to be exported.
A malicious instance is one which has become untrustworthy, either by hackers or malicious owners.

    Discussion: A malicious instance could lock their users in by refusing to allow them to export their toots.
    Mitigations: *If this is about the user being banned: Have a way to backup your toots every few days. If this is on iOS or Android, we can have the app save a backup automatically. On the web, we can store it into localStorage as well, but this is reliant on possibly malicious JS. If this is about a server that straight up does not provide the exporting functionality, there's not much recourse except to (1) warn the user when they first join (only available in non-server-controlled code like the apps) or (2) sue them if the user is in the EU.*

5. A malicious user sets up an account and imports someone else's toots, fraudulently claiming to be another person.

    Discussion: *I think this can already be done by malicious users by scraping existing toots.*
    Mitigations: ...

8. A user migrates their account to more than one other instances and/or accounts.

    Discussion: Perhaps the user forgot about the initial migration, or perhaps they wish to "fork" their account into multiple new accounts.
    Mitigations: *Allow the user to choose what toots to import... or just import all toots as usual, and then delete afterwards.*

**Comment by untitaker** [comment]
I have seen it claimed by at least one instance admin that they would want to refuse to import content because a post migration of a years-old account would be a lot of additional moderation burden at once. There are no abuse and failure scenarios here with regards to importing questionable content from one instance into the other, or even just content that is allowed on the old instance but not on the new one. And I'm not even necessarily talking about intentionally malicious uses of that feature, though it just gets worse there.

**Comment by brendonrapp** [comment]
Periodically exporting your history for backup purposes would mitigate the first 3 of these scenarios, and should be drilled into users as part of normal, "correct" usage of a federated platform.

> 1. An instance shuts down before toots can be exported.

Backups would ensure the user has their history saved up to the point of their most recent backup.

> 2. A malicious instance refuses to allow toots to be exported.

Backups would ensure the user has their history saved up to the point where the instance cut off the ability to export content. If exports had always been disallowed, that would have been an indication right from the beginning to not use that instance.

> 3. A malicious instance allows toots to be exported, but the toots are modified or corrupted in some way.

Backups taken prior to the malicious changes would ensure the user has the content saved in its original form.
